tools/flat-storage/src/commands.rs

```diff
@@ -400,23 +417,201 @@ impl FlatStorageCommand {
+            let prev_prev_hash = prev_header.prev_hash().clone();
+            let prev_height = prev_header.height();
+
+            let shard_layout = epoch_manager.get_shard_layout(&epoch_id)?;
+            shard_uid = epoch_manager.shard_id_to_uid(shard_id, epoch_id)?;
+
+            let trie =
+                runtime.get_trie_for_shard(shard_uid.shard_id(), &block_hash, state_root, false)?;
+
+            // Find all items which were changed after applying block `block_hash`
+            // and add them to delta.
+            // This is done by iterating over `DBCol::StateChanges` corresponding
+            // to given shard.
+            let mut prev_delta = FlatStateChanges::default();
+            for item in store.iter_prefix(DBCol::StateChanges, &block_hash.0) {
+                let (key, _) = item.unwrap();
+                let maybe_trie_key = &key[32..];
+                let maybe_account_id =
+                    near_primitives::trie_key::trie_key_parsers::parse_account_id_from_raw_key(
+                        maybe_trie_key,
+                    )?;
+                let maybe_trie_key = match maybe_account_id {
+                    Some(account_id) => {
+                        let account_shard_id = account_id_to_shard_id(&account_id, &shard_layout);
+                        if shard_id == account_shard_id {
+                            Some(maybe_trie_key)
+                        } else {
+                            None
+                        }
+                    }
+                    None => {
+                        assert!(maybe_trie_key.len() >= 8);
+                        let (trie_key, shard_uid_raw) =
+                            maybe_trie_key.split_at(maybe_trie_key.len() - 8);
+                        if shard_uid.to_bytes() == shard_uid_raw {
+                            Some(trie_key)
+                        } else {
+                            None
+                        }
+                    }
+                };
+
+                if let Some(trie_key) = maybe_trie_key {
+                    // Take *previous* value for the key from trie corresponding
+                    // to pre-state-root for this block.
+                    let prev_value = trie
+                        .get_optimized_ref(trie_key, near_store::KeyLookupMode::Trie)?
+                        .map(|value_ref| {
+                            near_primitives::state::FlatStateValue::Ref(value_ref.into_value_ref())
+                        });
+                    let value = store_helper::get_flat_state_value(&store, shard_uid, trie_key)?
+                        .map(|val| near_primitives::state::FlatStateValue::Ref(val.to_value_ref()));
+                    if prev_value != value {
+                        prev_delta.insert(trie_key.to_vec(), prev_value);
+                    }
+                }
+            }
+
+            // TODO(1.40): remove after removal of feature
```

Comment:
it is not clear which part of the rest of the code is subject to this TODO item. Maybe extract that code to a function and add the TODO on top of that function?